gio/src/input_stream.rs

```diff
@@ -227,25 +227,34 @@ impl<O: IsA<InputStream>> InputStreamExtManual for O {
         cancellable: Option<&C>,
         callback: Q,
     ) {
+        let main_context = glib::MainContext::ref_thread_default();
+        assert!(
+            main_context.is_owner(),
+            "Async operations only allowed if the thread is owning the MainContext"
+        );
+
         let cancellable = cancellable.map(|c| c.as_ref());
         let gcancellable = cancellable.to_glib_none();
-        let mut user_data: Box<Option<(Q, B)>> = Box::new(Some((callback, buffer)));
+        let mut user_data: Box<(glib::thread_guard::ThreadGuard<Q>, B)> =
+            Box::new((glib::thread_guard::ThreadGuard::new(callback), buffer));
```

Comment:
Also gets rid of some historical `Option` wrapping that is not needed anymore since quite a few Rust releases